src/audio/mfcc/mfcc_hifi3.c

```diff
@@ -0,0 +1,554 @@
+	 * remains zero.
+	 */
+	for (j = 0; j < state->prev_data_size; j++) {
+		AE_L16_XP(sample, in, buf_inc);
+		AE_S16_0_XP(sample, out, fft_inc);
+	}
+
+	/* Copy hop size of new data from circular buffer */
+	idx += state->prev_data_size;
+	in = (ae_int16 *)buf->r_ptr;
+	out = (ae_int16 *)&fft->fft_buf[idx].real;
+	set_circular_buf0(buf->addr, buf->end_addr);
+	for (j = 0; j < fft->fft_hop_size; j++) {
+		AE_L16_XC(sample, in, buf_inc);
+		AE_S16_0_XP(sample, out, fft_inc);
+	}
+
+	buf->s_avail -= fft->fft_hop_size;
+	buf->s_free += fft->fft_hop_size;
+	buf->r_ptr = in;
+
+	/* Copy for next time data back to overlap buffer */
+	idx = fft->fft_fill_start_idx + fft->fft_hop_size;
+	in = (ae_int16 *)&fft->fft_buf[idx].real;
+	out = (ae_int16 *)state->prev_data;
+	for (j = 0; j < state->prev_data_size; j++) {
+		AE_L16_XP(sample, in, fft_inc);
+		AE_S16_0_XP(sample, out, buf_inc);
+	}
+}
+
+#ifdef MFCC_NORMALIZE_FFT
+static int mfcc_normalize_fft_buffer(struct mfcc_state *state)
+{
+	struct mfcc_fft *fft = &state->fft;
+	ae_p16s *in = (ae_p16s *)&fft->fft_buf[fft->fft_fill_start_idx].real;
+	ae_int32x2 sample;
+	ae_int32x2 max = AE_ZERO32();
+	const int fft_inc = sizeof(fft->fft_buf[0]);
+	int shift;
+	int j;
+
+	for (j = 0; j < fft->fft_size; j++) {
+		/* load 16-bit data to middle of 32-bit container*/
+		AE_L16M_XU(sample, in, fft_inc);
+		max = AE_MAXABS32S(max, sample);
+	}
+	shift = AE_NSAZ32_L(max) - 8;/* 16 bit data */
+	shift = MAX(shift, 0);
+	shift = MIN(shift, MFCC_NORMALIZE_MAX_SHIFT);
+	return shift;
+}
+#endif
+
+static void mfcc_apply_window(struct mfcc_state *state, int input_shift)
+{
+	struct mfcc_fft *fft = &state->fft;
+	const int fft_inc = sizeof(fft->fft_buf[0]);
+	ae_int16 *win_in = (ae_int16 *)state->window;
+	const int win_inc = sizeof(ae_int16);
+	ae_int32x2 temp;
+	ae_int16x4 win;
+	int j;
+
+#if MFCC_FFT_BITS == 16
+	/* TODO: Use proper multiply and saturate function to make sure no overflows */
+	ae_int16 *fft_in = (ae_int16 *)&fft->fft_buf[fft->fft_fill_start_idx].real;
+	ae_int16x4 sample;
+
+	for (j = 0; j < fft->fft_size; j++) {
+		AE_L16_IP(sample, fft_in, 0);
+		AE_L16_XP(win, win_in, win_inc);
+		temp = AE_MULF16SS_00(sample, win);
+		temp = AE_SLAA32S(temp, input_shift);
+		sample = AE_ROUND16X4F32SASYM(temp, temp);
+		AE_S16_0_XP(sample, fft_in, fft_inc);
+	}
+#else
+	/* TODO: Use proper multiply and saturate function to make sure no overflows */
+	ae_int32 *fft_in = (ae_int32 *)&fft->fft_buf[fft->fft_fill_start_idx].real;
+	ae_int32x2 sample;
+
+	for (j = 0; j < fft->fft_size; j++) {
+		AE_L32_IP(sample, fft_in, 0);
+		AE_L16_XP(win, win_in, win_inc);
+		temp = AE_MULFP32X16X2RS_H(sample, win);
+		temp = AE_MULFP32X16X2RS_L(sample, win);
+		temp = AE_SLAA32S(temp, input_shift);
+		AE_S32_L_XP(temp, fft_in, fft_inc);
+	}
+#endif
+}
+
+/*
+ * The main processing function for MFCC
+ */
+
+static int mfcc_stft_process(const struct comp_dev *dev, struct mfcc_state *state)
```

Comment:
Would a mfcc_common.c be useful to keep common functions for implementations? I didn't check this function but looks like it's generic.